src/goto-instrument/contracts/dynamic-frames/dfcc_spec_functions.cpp

```diff
@@ -184,85 +236,78 @@ void dfcc_spec_functionst::generate_havoc_function(
+}
 
-  body.add(goto_programt::make_end_function());
-
-  goto_model.goto_functions.update();
+void dfcc_spec_functionst::to_spec_assigns_function(
+  const irep_idt &function_id,
+  std::size_t &nof_targets)
+{
+  auto &goto_function = goto_model.goto_functions.function_map.at(function_id);
 
-  std::set<irep_idt> no_body;
-  std::set<irep_idt> missing_function;
-  std::set<irep_idt> recursive_call;
-  std::set<irep_idt> not_enough_arguments;
-  utils.inline_function(
-    havoc_function_id,
-    no_body,
-    recursive_call,
-    missing_function,
-    not_enough_arguments);
-  INVARIANT(
-    no_body.empty(),
-    "no body warnings when inlining " + id2string(havoc_function_id));
-  INVARIANT(
-    missing_function.empty(),
-    "missing function warnings when inlining " + id2string(havoc_function_id));
-  INVARIANT(
-    recursive_call.empty(),
-    "recursive calls when inlining " + id2string(havoc_function_id));
-  INVARIANT(
-    not_enough_arguments.empty(),
-    "not enough arguments when inlining " + id2string(havoc_function_id));
+  // add write_set parameter
+  const exprt write_set_to_fill =
+    utils
+      .add_parameter(
+        function_id,
+        "__write_set_to_fill",
+        library.dfcc_type[dfcc_typet::WRITE_SET_PTR])
+      .symbol_expr();
 
-  utils.set_hide(havoc_function_id, true);
+  to_spec_assigns_instructions(
+    write_set_to_fill, goto_function.body, nof_targets);
 
   goto_model.goto_functions.update();
+
+  // instrument for side-effects checking
+  std::set<irep_idt> function_pointer_contracts;
+  instrument.instrument_function(function_id, function_pointer_contracts);
+  INVARIANT(
+    function_pointer_contracts.size() == 0,
+    "discovered function pointer contracts unexpectedly");
+  utils.set_hide(function_id, true);
 }
 
-void dfcc_spec_functionst::to_spec_assigns_function(
-  const irep_idt &function_id,
+void dfcc_spec_functionst::to_spec_assigns_instructions(
+  const exprt &write_set_to_fill,
+  goto_programt &program,
   std::size_t &nof_targets)
 {
   // counts the number of calls to built-ins to get an over approximation
   // of the size of the set
   std::size_t next_idx = 0;
 
-  auto &goto_function = goto_model.goto_functions.function_map.at(function_id);
-
-  // add write_set parameter
-  auto &set_symbol = utils.add_parameter(
-    function_id,
-    "__write_set_to_fill",
-    library.dfcc_type[dfcc_typet::WRITE_SET_PTR]);
-
   // rewrite calls
-  Forall_goto_program_instructions(ins_it, goto_function.body)
+  Forall_goto_program_instructions(ins_it, program)
   {
     if(ins_it->is_function_call())
     {
       if(ins_it->call_function().id() != ID_symbol)
       {
         throw invalid_source_file_exceptiont(
           "Function pointer call '" +
-            from_expr(ns, function_id, ins_it->call_function()) +
-            "' in function '" + id2string(function_id) + "' is not supported",
+            from_expr(ns, "", ins_it->call_function()) +
```

Comment:
Without an identifier we cannot determine the mode, i.e., the language. It might then be better to fully fall back to `format_to_string`, which is language agnostic.

Reply:
Added the language mode as extra parameter